### ISSUE TITLE
A function that is wrapped with tf.function gives wrong gradients on loss functions that are themselves wrapped by the same decorator

### ISSUE THREAD

**Issue by borundev** [tensorflow/tensorflow#40984]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
No
- TensorFlow installed from (source or binary):
Binary
- TensorFlow version (use command below):
2.2.0
- Python version:
3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

I have two versions of loss function, one decorated with tf.function and the other not,  and I have two functions that compute the gradients using these two loss functions, one decorate with tf.function and the other not . When I use the gradient function that is not decorate I get the same results for both the loss function. However, when I use the gradient function that is decorated I get different results and in particular I get a wrong result for the loss function that is decorated.

**Describe the expected behavior**

I would expect all of the above to give the same result.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

https://colab.research.google.com/drive/1erNOChp355OTiCwWiizPdjA1OT1s3HYV?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/6a0afaa83051077734f18b5032f8ef1c/40984.ipynb). Thanks!

**Comment by nikitamaia** [comment]
Hi @borundev, you mentioned that when you use the gradient function that is decorated, you get the wrong result for the loss function that is decorated.

I'm running your colab and seeing that the different result is for the loss function that is *not* decorated. Can you double check and let me know if I'm reading this correctly?

**Comment by borundev** [comment]
Hi @nikitamaia you are right. I must have mistyped it here. When I asked this on [stackoverflow ](https://stackoverflow.com/questions/62428590/gradienttape-gives-different-gradients-depending-on-loss-function-being-decorate) I said 

> Thus, when my outer function is decorated I only get the correct answer from the inner function that is decorated as well. I was under the impression that decorating the outer one (which is the training loop in many applications) is sufficient but here we see its not. 

I apologize for the confusion.

**Comment by nikitamaia** [comment]
No worries! 
Using from_logits=True is more numerically stable (see [note in docs here](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/losses.py#L554)) and recommended. If you do that and also remove the softmax from your output layer you should see the same result in all cases.

**Comment by borundev** [comment]
@nikitamaia So this is not considered a bug?.

**Comment by nikitamaia** [comment]
Hi @borundev you are correct, it is a bug since the behavior is surprising and from_logits=False does not error out. And I believe this is a duplicate of issue #32895. I will close this issue now so we can track this bug in one place, but feel free to reopen if you don't think it's a duplicate.
